File: app/src/main/java/no/nav/tilbakemeldingsmottak/rest/serviceklage/service/HentSkjemaService.java
package no.nav.tilbakemeldingsmottak.rest.serviceklage.service;

import static no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.ServiceklageConstants.BRUKER_IKKE_BEDT_OM_SVAR_ANSWER;
import static no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.ServiceklageConstants.ENHETSNUMMER_BEHANDLENDE;
import static no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.ServiceklageConstants.ENHETSNUMMER_PAAKLAGET;
import static no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.ServiceklageConstants.FREMMET_DATO;
import static no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.ServiceklageConstants.INNSENDER;
import static no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.ServiceklageConstants.KANAL;
import static no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.ServiceklageConstants.SVARMETODE;
import static no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.ServiceklageConstants.SVARMETODE_UTDYPNING;
import static no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.ServiceklageConstants.SVAR_IKKE_NOEDVENDIG;
import static no.nav.tilbakemeldingsmottak.util.SkjemaUtils.getQuestionById;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import lombok.extern.slf4j.Slf4j;
import no.nav.tilbakemeldingsmottak.consumer.norg2.Enhet;
import no.nav.tilbakemeldingsmottak.consumer.norg2.Norg2Consumer;
import no.nav.tilbakemeldingsmottak.exceptions.SkjemaConstructionException;
import no.nav.tilbakemeldingsmottak.exceptions.SkjemaSerializationException;
import no.nav.tilbakemeldingsmottak.repository.ServiceklageRepository;
import no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.Answer;
import no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.DefaultAnswers;
import no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.HentSkjemaResponse;
import no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.Serviceklage;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.Resource;
import org.springframework.stereotype.Service;
import org.springframework.util.StreamUtils;

import java.io.IOException;
import java.nio.charset.Charset;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@Slf4j
public class HentSkjemaService {

    private final ServiceklageRepository serviceklageRepository;
    private final Norg2Consumer norg2Consumer;
    private final String classpathSkjema;
    private ObjectMapper mapper = new ObjectMapper(new YAMLFactory());

    private static final String MESSAGE = "Feltet ble fylt ut under registrering av serviceklage";
    private static final String NEDLAGT = "nedlagt";
    private static final String ANNET = "Annet";
    private static final String SCHEMA_PATH = "classpath:schema/schema.yaml";
    private static final Charset CHARSET = Charset.forName("utf-8");

    public HentSkjemaService(@Value(SCHEMA_PATH) Resource schema,
                             ServiceklageRepository serviceklageRepository,
                             Norg2Consumer norg2Consumer) throws IOException {
        mapper.findAndRegisterModules();
        this.classpathSkjema = StreamUtils.copyToString(schema.getInputStream(), CHARSET);
        this.serviceklageRepository = serviceklageRepository;
        this.norg2Consumer = norg2Consumer;
    }

    public HentSkjemaResponse hentSkjema(String journalpostId) {
        HentSkjemaResponse response = readSkjema();

        List<Answer> enheter = hentEnheter();

        getQuestionById(response.getQuestions(), ENHETSNUMMER_PAAKLAGET)
                .orElseThrow(() -> new SkjemaConstructionException("Finner ikke spørsmål med id=" + ENHETSNUMMER_PAAKLAGET))
                .setAnswers(enheter);

        getQuestionById(response.getQuestions(), ENHETSNUMMER_BEHANDLENDE)
                .orElseThrow(() -> new SkjemaConstructionException("Finner ikke spørsmål med id=" + ENHETSNUMMER_BEHANDLENDE))
                .setAnswers(enheter);

        Serviceklage serviceklage = serviceklageRepository.findByJournalpostId(journalpostId);
        if (serviceklage != null) {
            response.setDefaultAnswers(DefaultAnswers.builder()
                    .message(MESSAGE)
                    .answers(mapDefaultAnswers(serviceklage))
                    .build());
        }

        return response;
    }

    public HentSkjemaResponse readSkjema() {
        try {
            return mapper.readValue(classpathSkjema, HentSkjemaResponse.class);
        } catch (Exception e) {
            throw new SkjemaSerializationException("Feil under serialisering av skjema");
        }
    }

    private Map<String, String> mapDefaultAnswers(Serviceklage serviceklage) {
        Map<String, String> defaultAnswers = new HashMap<>();
        if (serviceklage.getFremmetDato() != null) {
            defaultAnswers.put(FREMMET_DATO, serviceklage.getFremmetDato().toString());
        }
        if (serviceklage.getInnsender() != null) {
            defaultAnswers.put(INNSENDER, serviceklage.getInnsender());
        }
        if (serviceklage.getKanal() != null) {
            defaultAnswers.put(KANAL, serviceklage.getKanal());
        }
        if (serviceklage.getSvarmetode() != null) {
            defaultAnswers.put(SVARMETODE, serviceklage.getSvarmetode());
        }
        if (serviceklage.getSvarmetodeUtdypning() != null) {
            if (BRUKER_IKKE_BEDT_OM_SVAR_ANSWER.equals(serviceklage.getSvarmetodeUtdypning())) {
                defaultAnswers.put(SVAR_IKKE_NOEDVENDIG, serviceklage.getSvarmetodeUtdypning());
            } else {
                defaultAnswers.put(SVAR_IKKE_NOEDVENDIG, ANNET);
                defaultAnswers.put(SVARMETODE_UTDYPNING, serviceklage.getSvarmetodeUtdypning());
            }
        }
        return defaultAnswers;
    }

    private List<Answer> hentEnheter() {
        List<Enhet> enheter = norg2Consumer.hentEnheter();
        return enheter.stream()
                .filter(e -> !e.getStatus().equalsIgnoreCase(NEDLAGT))
                .map(e -> Answer.builder().answer(e.getNavn() + " - " + e.getEnhetNr()).build())
                .collect(Collectors.toList());
    }

}

File: app/src/main/kotlin/no/nav/tilbakemeldingsmottak/rest/serviceklage/service/HentSkjemaService.kt
package no.nav.tilbakemeldingsmottak.rest.serviceklage.service

import no.nav.tilbakemeldingsmottak.consumer.norg2.Norg2Consumer
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.BRUKER_IKKE_BEDT_OM_SVAR_ANSWER
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.ENHETSNUMMER_BEHANDLENDE
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.ENHETSNUMMER_PAAKLAGET
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.FREMMET_DATO
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.INNSENDER
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.KANAL
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.SVARMETODE
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.SVARMETODE_UTDYPNING
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.SVAR_IKKE_NOEDVENDIG
import no.nav.tilbakemeldingsmottak.domain.models.Serviceklage
import no.nav.tilbakemeldingsmottak.exceptions.ServerErrorException
import no.nav.tilbakemeldingsmottak.model.Answer
import no.nav.tilbakemeldingsmottak.model.DefaultAnswers
import no.nav.tilbakemeldingsmottak.model.HentSkjemaResponse
import no.nav.tilbakemeldingsmottak.repository.ServiceklageRepository
import no.nav.tilbakemeldingsmottak.util.SkjemaUtils.Companion.getQuestionById
import no.nav.tilbakemeldingsmottak.util.SkjemaUtils.Companion.updateQuestionsById
import org.springframework.beans.factory.annotation.Value
import org.springframework.core.io.Resource
import org.springframework.stereotype.Service
import org.springframework.util.StreamUtils
import tools.jackson.dataformat.yaml.YAMLMapper
import tools.jackson.module.kotlin.kotlinModule
import java.nio.charset.StandardCharsets

@Service
class HentSkjemaService(
    @Value(SCHEMA_PATH) schema: Resource,
    private val serviceklageRepository: ServiceklageRepository,
    private val norg2Consumer: Norg2Consumer
) {
    companion object {
        private const val SCHEMA_PATH = "classpath:schema/schema.yaml"
    }

    private var classpathSkjema: String = ""
    private val MESSAGE = "Feltet ble fylt ut under registrering av serviceklage"
    private val NEDLAGT = "nedlagt"
    private val ANNET = "Annet"
    private val CHARSET = StandardCharsets.UTF_8


    private val yamlMapper = YAMLMapper.builder()
        .addModule(kotlinModule())     // register Kotlin support
        .build()

    init {
        classpathSkjema = StreamUtils.copyToString(schema.inputStream, CHARSET)
    }

    // Get questions/answers from schema.yaml and update the answers with "enheter" from norg2 og default answers from "serviceklage"
    fun hentSkjema(journalpostId: String): HentSkjemaResponse {
        var response = readSkjema()
        val enheter = hentEnheter()

        response = setAnswersOnQuestions(response, ENHETSNUMMER_PAAKLAGET, enheter)
        response = setAnswersOnQuestions(response, ENHETSNUMMER_BEHANDLENDE, enheter)

        serviceklageRepository.findByJournalpostId(journalpostId)?.let { serviceklage ->
            response = response.copy(
                defaultAnswers = DefaultAnswers(
                    message = MESSAGE,
                    answers = mapDefaultAnswers(serviceklage)
                )
            )
        }

        return response
    }

    // Spørsmålet kan være duplisert inn i flere svargrener, så alle forekomstene må få enhetene
    private fun setAnswersOnQuestions(
        response: HentSkjemaResponse,
        id: String,
        answers: List<Answer>
    ): HentSkjemaResponse {
        val questions = response.questions
        if (questions == null || getQuestionById(questions, id) == null) {
            throw ServerErrorException("Finner ikke spørsmål med id=$id")
        }
        return response.copy(questions = updateQuestionsById(questions, id) { it.copy(answers = answers) })
    }


    fun readSkjema(): HentSkjemaResponse {
        return try {
            yamlMapper.readValue(classpathSkjema, HentSkjemaResponse::class.java)
        } catch (e: Exception) {
            throw ServerErrorException("Feil under serialisering av skjema", e)
        }
    }

    private fun mapDefaultAnswers(serviceklage: Serviceklage): Map<String, String> {
        return hashMapOf<String, String>().apply {
            serviceklage.fremmetDato?.let { put(FREMMET_DATO, it.toString()) }
            serviceklage.innsender?.let { put(INNSENDER, it) }
            serviceklage.kanal?.let { put(KANAL, it) }
            serviceklage.svarmetode?.let { put(SVARMETODE, it) }
            serviceklage.svarmetodeUtdypning?.let {
                when (it) {
                    BRUKER_IKKE_BEDT_OM_SVAR_ANSWER -> put(SVAR_IKKE_NOEDVENDIG, it)
                    else -> {
                        put(SVAR_IKKE_NOEDVENDIG, ANNET)
                        put(SVARMETODE_UTDYPNING, it)
                    }
                }
            }
        }
    }

    private fun hentEnheter(): List<Answer> {
        val enheter = norg2Consumer.hentEnheter()
        return enheter
            .filter { !it.status.equals(NEDLAGT, ignoreCase = true) }
            .map { Answer(answer = "${it.navn} - ${it.enhetNr}") }
    }
}
